sql: optimize locking behavior of delete portion of UPDATEs and UPSERTs

yuzefovich opened the pull request:
**execbuilder: improve tracing tests for UPDATEs and UPSERTs**

The updated tests will be used to show-case changes to lock acquisitions
for UPDATEs and UPSERTs with buffered writes. Also address a couple
TODOs by adding tests for "blind" Upsert and Delete fast-paths.

**sql: optimize locking behavior of delete portion of UPDATEs and UPSERTs**

This commit applies the same lock acquisition optimizations to `Del`
commands that are issued as part of UPDATEs and UPSERTs as we've already
applied for DELETEs. Namely, we don't need to performing locking Dels
when
- the initial scan has already acquired locks on the keys being deleted
- when deleting from non-unique secondary indexes.

Informs: #139106.
Informs: #139160.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142250)
<!-- Reviewable:end -->

yuzefovich committed:
execbuilder: improve tracing tests for UPDATEs and UPSERTs

The updated tests will be used to show-case changes to lock acquisitions
for UPDATEs and UPSERTs with buffered writes. Also address a couple
TODOs by adding tests for "blind" Upsert and Delete fast-paths.

Release note: None

yuzefovich committed:
sql: optimize locking behavior of delete portion of UPDATEs and UPSERTs

This commit applies the same lock acquisition optimizations to `Del`
commands that are issued as part of UPDATEs and UPSERTs as we've already
applied for DELETEs. Namely, we don't need to performing locking Dels
when
- the initial scan has already acquired locks on the keys being deleted
- when deleting from non-unique secondary indexes.

Release note: None

yuzefovich commented:
I'm still thinking about the "update" portion of UPDATEs and UPSERTs but figured that can go into a separate PR. RFAL.

yuzefovich commented:
If folks have bandwidth, #143247 includes the "update" portion of UPDATEs and is RFAL. Keeping that PR in draft only until this merges.

stevendanna reviewed (approved):
> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142250)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @michae2)

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/142250#-OLravxd-SxdmvwONkXy:-OLravxd-SxdmvwONkXz:ba3fsrp) (no related file):*
I really like the tests you added here. I always find the trace-based tests so informative.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice!

> Reviewed 5 of 5 files at r1, 23 of 23 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142250#-:-OLu_Ap098wTnr6jGwou:b-8gk8uj)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @yuzefovich)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/update` line 1335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142250#-OLuOXIKCXUtRaBBSATf:-OLuOXIKCXUtRaBBSATg:b-tpn18a) ([raw file](https://github.com/cockroachdb/cockroach/blob/82d1deeaf71847487edb6547a242cd2f410f46c3/pkg/sql/opt/exec/execbuilder/testdata/update#L1335)):*
> ```
>   v INT,
>   INDEX (i) STORING (u),
>   UNIQUE INDEX (u) STORING (i),
> ```

Might also be interesting to add:
* a unique index on a non-nullable column that is not storing anything
* indexes that end up with multiple column families
___
*[`pkg/sql/row/updater.go` line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142250#-OLuWtnC9cv8MfWP7G93:-OLuWtnC9cv8MfWP7G94:blinnsj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5da745f48eaba7dfe36f128f1e7a87336510d0f/pkg/sql/row/updater.go#L186)):*
> ```Go
> 	}
> 	if len(lockedIndexes) > 1 && !primaryLocked {
> 		panic(errors.AssertionFailedf("locked at least two secondary indexes in the initial scan: %v", lockedIndexes))
> ```

I think in https://github.com/cockroachdb/cockroach/pull/141404 you had the equivalent assertion in execbuilder behind `buildutil.CrdbTestBuild`. Should the same be true for this assertion? (If we fail it, I believe nothing breaks, we just perform more locking than we strictly need to.)

(Rather than enforce this one-secondary-index-is-locked invariant, would it be simpler to relax the invariant and not assume anything about which secondary indexes are already locked?)


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142250#-:-OLvAnRT4YtpFiAVobQp:b-tlx7ko)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @michae2)

___
*[`pkg/sql/row/updater.go` line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142250#-OLuWtnC9cv8MfWP7G93:-OLv6ncW1MO7ccBNQmLk:b-kvbwdu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5da745f48eaba7dfe36f128f1e7a87336510d0f/pkg/sql/row/updater.go#L186)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I think in https://github.com/cockroachdb/cockroach/pull/141404 you had the equivalent assertion in execbuilder behind `buildutil.CrdbTestBuild`. Should the same be true for this assertion? (If we fail it, I believe nothing breaks, we just perform more locking than we strictly need to.)

(Rather than enforce this one-secondary-index-is-locked invariant, would it be simpler to relax the invariant and not assume anything about which secondary indexes are already locked?)
</blockquote></details>

Yeah, good point. Switched to test-only assertion with some comments.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/update` line 1335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142250#-OLuOXIKCXUtRaBBSATf:-OLv9J5E-JPJzD47CIrL:bp3ryky) ([raw file](https://github.com/cockroachdb/cockroach/blob/82d1deeaf71847487edb6547a242cd2f410f46c3/pkg/sql/opt/exec/execbuilder/testdata/update#L1335)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Might also be interesting to add:
* a unique index on a non-nullable column that is not storing anything
* indexes that end up with multiple column families
</blockquote></details>

Thanks. I included unique secondary on non-nullable column here. I didn't include any more column families since we have good multiple column family coverage in `secondary_index_column_families` file (which I'm touching in the following PR). The reason why I added a second column family in this case is to verify the `Del`s from the primary index (which are now `locking` in some cases, as of this PR).


<!-- Sent from Reviewable.io -->
